rustls/src/key_schedule.rs

```diff
@@ -177,6 +180,14 @@ impl KeyScheduleHandshake {
         secret
     }
 
+    pub fn server_ech_confirmation_secret(&mut self, hs_hash: &Digest) -> PayloadU8 {
```

Comment:
The idea behind the the *ring* HKDF API is that you would do something like this:

```
pub struct Iv([u8; aead::NONCE_LEN]);
struct IvLen;

impl hkdf::KeyType for IvLen {
    fn len(&self) -> usize {
        aead::NONCE_LEN
    }
}
```

s/`Iv`/`EchConfirmationSecret`/.

Reply:
I don't think this code will live on in the final implementation, given https://github.com/tlswg/draft-ietf-tls-esni/pull/420/files, so I just did the easiest thing for interop testing in draft-10.

Reply:
Also, I would say all the existing uses of `PayloadU8Len` are unfortunate, so I wouldn't copy them. `derive_ticket_psk` in particular should not be using `Vec<u8>` as its return type, but instead it should use a stronger key type.

Ideally `export_keying_material` would have a more sophisticated API as well, more like *ring*'s HKDF API; I think the desire to avoid *ring* types in Rust's public API is what makes that difficult.

OTOH, it probably it makes more sense for `server_ech_confirmation_secret` to compute the return the entire `Random` including the 24 random bytes and the 8 derived bytes.

Reply:
To add to my above comment, I think the draft-11 ECH confirmation will be able to happen entirely in client/hs.rs. The draft-10 design requires changes to tls13.rs and key_schedule.rs, but I don't think those will be needed in the future.